File: problems/problem_4.py
from utils.functions import is_palindrome


class Problem4:
    TITLE = 'Largest palindrome product'
    DESCRIPTION = 'https://projecteuler.net/problem=4'

    def answer(self) -> int:
        answer = float('-inf')
        for num1 in reversed(range(100, 999)):
            for num2 in reversed(range(100, 999)):
                if is_palindrome(num1*num2):
                    answer = max(answer, num1*num2)

        return answer

File: rust/src/problems/problem_4.rs
pub mod problem_4 {
    pub const TITLE: &str = "Largest palindrome product";
    pub const DESCRIPTION: &str = "https://projecteuler.net/problem=4";

    pub fn answer() -> u32 {
        use std::cmp;
        use crate::utils::functions::functions_u32::is_palindrome;

        let mut answer: u32 = 0;
        for num_1 in (100..999).rev() {
            for num_2 in (100..999).rev() {
                if is_palindrome(num_1 * num_2) {
                    answer = cmp::max(answer, num_1 * num_2);
                }
            }
        }
        answer
    }

    pub fn pretty_print_solution() {
        use std::time::Instant;

        let now = Instant::now();
        let solution = answer();
        let elapsed = now.elapsed();

        println!("Problem {} - {}", 4, crate::problems::problem_4::problem_4::TITLE);
        println!("Description: {}", crate::problems::problem_4::problem_4::DESCRIPTION);
        println!("Solution: {}", solution);
        println!("Time: {:.2?}", elapsed);
    }

}